tests/phpunit/tests/option/option.php

```diff
@@ -359,16 +359,103 @@ public function test_option_autoloading( $name, $autoload_value, $expected ) {
 	 */
 	public function data_option_autoloading() {
 		return array(
-			array( 'autoload_yes', 'yes', 'yes' ),
-			array( 'autoload_true', true, 'yes' ),
-			array( 'autoload_string', 'foo', 'yes' ),
-			array( 'autoload_int', 123456, 'yes' ),
-			array( 'autoload_array', array(), 'yes' ),
-			array( 'autoload_no', 'no', 'no' ),
-			array( 'autoload_false', false, 'no' ),
+			array( 'autoload_yes', 'yes', 'on' ),
+			array( 'autoload_true', true, 'on' ),
+			array( 'autoload_string', 'foo', 'on' ),
+			array( 'autoload_int', 123456, 'on' ),
+			array( 'autoload_array', array(), 'on' ),
+			array( 'autoload_no', 'no', 'off' ),
+			array( 'autoload_false', false, 'off' ),
+			array( 'autoload_null', null, 'auto' ),
 		);
 	}
 
+	/**
+	 *
+	 * @ticket 42441
+	 * @dataProvider data_option_autoloading_large_option
+	 *
+	 * @covers ::update_option
+	 */
+	public function test_update_option_autoloading_large_option( $autoload, $expected ) {
+		global $wpdb;
+		$name = 'foo';
+		add_option( $name, 'bar' );
+		add_filter( 'wp_max_autoloaded_option_size', array( $this, 'filter_max_option_size' ) );
+		$value = file( DIR_TESTDATA . '/formatting/entities.txt' );
+		$added = update_option( $name, $value, $autoload );
+		$this->assertTrue( $added );
+
+		$actual = $wpdb->get_row( $wpdb->prepare( "SELECT autoload FROM $wpdb->options WHERE option_name = %s LIMIT 1", $name ) );
+		$this->assertSame( $expected, $actual->autoload );
+	}
+
+	public function data_option_autoloading_large_option() {
+		return array(
+			'on'   => array(
+				'autoload' => 'on',
+				'expected' => 'on',
+			),
+			'off'  => array(
+				'autoload' => 'off',
+				'expected' => 'off',
+			),
+			'yes'   => array(
+				'autoload' => 'yes',
+				'expected' => 'on',
+			),
+			'true'  => array(
+				'autoload' => true,
+				'expected' => 'on',
+			),
+			'no'    => array(
+				'autoload' => 'no',
+				'expected' => 'off',
+			),
+			'false' => array(
+				'autoload' => false,
+				'expected' => 'off',
+			),
+			'null'  => array(
+				'autoload' => null,
+				'expected' => 'auto-off',
+			),
+			'autoload_int'  => array(
+				'autoload' => '1223',
+				'expected' => 'on',
+			),
+			'autoload_array'  => array(
+				'autoload' => array(),
+				'expected' => 'on',
+			),
+			'stdClass'  => array(
+				'autoload' => new stdClass(),
+				'expected' => 'on',
+			),
+		);
+	}
+
+	public function filter_max_option_size( $current ) {
+		return 1000;
+	}
+
+	/**
+	 *
+	 * @ticket 42441
+	 *
+	 * @covers ::update_option
+	 */
+	public function test_update_option_autoloading_small_option_auto() {
+		global $wpdb;
+		$name = 'foo';
+		add_option( $name, 'bar', '', 'auto' );
+		$added = update_option( $name, 'small_option_data' );
+		$this->assertTrue( $added );
+
+		$actual = $wpdb->get_row( $wpdb->prepare( "SELECT autoload FROM $wpdb->options WHERE option_name = %s LIMIT 1", $name ) );
+		$this->assertSame( 'on', $actual->autoload );
+	}
```

Comment:
This test doesn't look right. It should assert that WP core correctly assigns `auto` to a small option without an explicit value. But you're providing `auto` as an explicit value, which shouldn't be provided, and thus results in `on`.
```suggestion
	public function test_update_option_autoloading_small_option_auto() {
		global $wpdb;

		$name = 'foo';
		add_option( $name, 'bar' );
		$updated = update_option( $name, 'small_option_data' );
		$this->assertTrue( $updated );

		$actual = $wpdb->get_row( $wpdb->prepare( "SELECT autoload FROM $wpdb->options WHERE option_name = %s LIMIT 1", $name ) );
		$this->assertSame( 'auto', $actual->autoload );
	}
```